staging/src/k8s.io/api/admissionregistration/v1/types.go

```diff
@@ -307,6 +307,20 @@ type ValidatingWebhook struct {
 	// include any versions known to the API Server, calls to the webhook will fail
 	// and be subject to the failure policy.
 	AdmissionReviewVersions []string `json:"admissionReviewVersions" protobuf:"bytes,8,rep,name=admissionReviewVersions"`
+
+	// MatchConditions is a list of conditions on the AdmissionRequest that must be met
+	// for a request to be sent to this webhook. All conditions must be met for the request to be matched.
+	// An empty list of matchConditions matches all requests. If there is an error evaluating the condition,
+	// the error is ignored and the condition is considered matched.
```

Comment:
> what does the CEL authorizer do for errors? can a CEL expression use the authorizer function and "catch" an error and turn it into a boolean return?

CEL calls to the authorizer `.check()` functions may evaluate to runtime errors, which cannot be caught.

We could change this.  The alternative would be for `.check()` to always return a `Decision` object, even if the underlying Go Authorize() call returns an error.  The `Decision` could hold the error and offer an API like:

`.isError() bool`
`.isAllowed() bool` (raises an error if the `isError()` is true, otherwise guaranteed to not raise an error)
`.reason() string` (probably should always return a string, returning the error in a formatted string if `isError()` is true)

Thoughts?

Reply:
done with @tallclair doc commit

Reply:
@liggitt @tallclair 

One thing occurred to me around how these can fail... exceeding compute budget.

We've been assuming the only source of failure is the code itself. This means that taking an approach like "any no match => no match, regardless of there being other errors" de-risks the potential for beyond-scope enforcement caused by assuming error => match, but can failure also be induced by exceeding the CEL budget? If so, could a user purposefully write an expensive CEL statement to force failure, forcing a request to go to the webhook?

It's tempting to say the malicious author case is moot, since only trusted people should have the ability to modify webhooks, but since we are considering things like cloud providers injecting these statements, shared ownership is implicit, and this could be a way for one owner to veto the others.

Reply:
that's a really good point... if we are limiting overall CEL cost (and I hope we are), I agree that an early expensive condition that eats the whole budget shouldn't circumvent a later condition that would have returned false had it had a chance to run

Reply:
One possible way around this is to evaluate statements in-order, and then a cloud provider can guarantee their (light-weight) condition is always checked by inserting it at the beginning of the list (there are some nuances around a user modifying or removing the condition, but those were always there).

Reply:
If we combined the insert at the beginning of the list with a VAP that required it at the beginning of the list for all webhooks it would also prevent user modification/removal. The current implementation does retain order but if this becomes a critical piece of the implementation we may want some safeguards around ordering.